jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/EditLambdaDialog.kt

```diff
@@ -0,0 +1,363 @@
+        view.timeout.text = timeout.toString()
+        view.description.text = description
+        view.envVars.envVars = envVariables
+
+        view.sourceBucket.populateValues {
+            val activeRegionId = ProjectAccountSettingsManager.getInstance(project).activeRegion.id
+            s3Client.listBuckets().buckets()
+                .asSequence()
+                .mapNotNull { it?.name() }
+                .filter { s3Client.regionForBucket(it) == activeRegionId }
+                .sortedWith(String.CASE_INSENSITIVE_ORDER)
+                .toList()
+        }
+
+        view.createBucket.addActionListener {
+            val bucket = Messages.showInputDialog(
+                message("lambda.upload.create_s3_dialog.input"),
+                message("lambda.upload.create_s3_dialog.title"),
+                AwsIcons.Logos.S3_LARGE
+            )
+            bucket?.run {
+                view.sourceBucket.addAndSelectValue {
+                    s3Client.createBucket { request -> request.bucket(bucket) }
+                    bucket
+                }
+            }
+        }
+
+        view.runtime.populateValues(selected = runtime) { Runtime.knownValues() }
+
+        view.iamRole.populateValues(selected = role) {
+            iamClient.listRolesFilter { it.assumeRolePolicyDocument().contains(LAMBDA_PRINCIPAL) }
+                .map { IamRole(name = it.roleName(), arn = it.arn()) }
+                .sortedWith(Comparator.comparing<IamRole, String>(Function { it.name }, String.CASE_INSENSITIVE_ORDER))
+                .toList()
+        }
+
+        view.createRole.addActionListener {
+            val iamRoleDialog = CreateIamRoleDialog(
+                project = project,
+                iamClient = iamClient,
+                parent = view.content,
+                defaultAssumeRolePolicyDocument = DEFAULT_ASSUME_ROLE_POLICY,
+                defaultPolicyDocument = DEFAULT_POLICY
+            )
+            if (iamRoleDialog.showAndGet()) {
+                iamRoleDialog.iamRole?.let { newRole -> view.iamRole.addAndSelectValue { newRole } }
+            }
+        }
+    }
+
+    override fun createCenterPanel(): JComponent? = view.content
+
+    override fun getPreferredFocusedComponent(): JComponent? = view.name
+
+    override fun doValidate(): ValidationInfo? {
+        val validateSettings = validator.validateSettings(view)
+        updateAction.isEnabled = validateSettings == null // We can update, but not deploy
+
+        return validateSettings ?: if (validateDeploySettings) validator.validateDeploymentSettings(project, view) else null
+    }
+
+    override fun getOKAction(): Action = deployAction
+
+    override fun createActions(): Array<Action> {
+        val actions = mutableListOf<Action>()
+        actions.add(okAction)
+        if (isUpdate) {
+            actions.add(updateAction)
+        }
+        actions.add(cancelAction)
+
+        return actions.toTypedArray()
+    }
+
+    override fun doOKAction() {
```

Comment:
Difference between `doOKAction` and `getOKAction` ?

Reply:
`getOKAction` returns the actual Action class, `doOKAction` closes the window (by default)

```java
  protected void doOKAction() {
    if (getOKAction().isEnabled()) {
      close(OK_EXIT_CODE);
    }
  }
```

Reply:
I guess my question was more around the handling of closure. So because we have custom 'update settings' and 'deploy lambda' buttons/actions we're explicitly removing the default close logic